SomeSortofCommerce,YouSay/app/src/main/java/com/jonathanlieblich/somesortofcommerceyousay/ProductViewActivity.java

```diff
@@ -0,0 +1,126 @@
+            public void onClick(View view) {
+                switch(view.getId()) {
+                    case R.id.fab:
+                        Intent intent = new Intent(ProductViewActivity.this, ShoppingCartActivity.class);
+                        startActivity(intent);
+                        break;
+                }
+            }
+        };
+        fab.setOnClickListener(onClickListener);
+    }
+
+    private class AsyncProductViewTask extends AsyncTask<Void, Integer, List<Product>> {
+
+        @Override
+        protected List<Product> doInBackground(Void... voids) {
+            return ProductStorageHelper.getInstance(getApplicationContext()).getProductList();
+        }
+
+        @Override
+        protected void onPostExecute(List<Product> products) {
+            super.onPostExecute(products);
+            mAdapter = new ProductViewAdapter(products);
+            mainRecycler = (RecyclerView)findViewById(R.id.product_recycler_view);
+            LinearLayoutManager layoutManager = new LinearLayoutManager(getApplicationContext(),
+                    LinearLayoutManager.HORIZONTAL, false);
+            mainRecycler.setLayoutManager(layoutManager);
+            mainRecycler.setAdapter(mAdapter);
+        }
+    }
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        getMenuInflater().inflate(R.menu.menu_product_view, menu);
+
+        SearchManager searchmanager = (SearchManager)getSystemService(Context.SEARCH_SERVICE);
+        SearchView searchView = (SearchView)menu.findItem(R.id.action_search).getActionView();
+        ComponentName componentName = new ComponentName(this, ProductViewActivity.class);
+        searchView.setSearchableInfo(searchmanager.getSearchableInfo(componentName));
+
+        return true;
+    }
+
+    @Override
+    protected void onNewIntent(Intent intent) {
+        setIntent(intent);
+        handleIntent(intent);
+    }
+
+    private void handleIntent(Intent intent) {
+        if(Intent.ACTION_SEARCH.equals(intent.getAction())) {
+            String query = intent.getStringExtra(SearchManager.QUERY);
+
+            AsyncSearchTask task = new AsyncSearchTask();
+            task.execute(query);
+        }
+    }
+
+    private class AsyncSearchTask extends AsyncTask<String, Void, List<Product>> {
+
+        @Override
+        protected List<Product> doInBackground(String... strings) {
+            return ProductStorageHelper.getInstance(getApplicationContext()).searchByName(strings[0]);
+        }
+
+        @Override
+        protected void onPostExecute(List<Product> products) {
+            super.onPostExecute(products);
+            if(products.size() < 1) {
+                Toast.makeText(ProductViewActivity.this, "No items matched your search", Toast.LENGTH_SHORT).show();
+                products = ProductStorageHelper.getInstance(getApplicationContext()).getProductList();
+            }
+            mainRecycler.setAdapter(new ProductViewAdapter(products));
```

Comment:
Don't create a whole new adapter just to update the RecyclerView! This is very inefficient. Instead, update the underlying of `Product`s that the adapter uses and call the `notifyDataSetChanged()` method.